### PR TITLE
Generate language-specific compile macro files

### PR DESCRIPTION
**Pull request by brondani**
## Fixes
- https://github.com/Open-CMSIS-Pack/cmsis-toolbox/issues/625
- https://github.com/Open-CMSIS-Pack/vscode-cmsis-solution/issues/335

Depends on https://github.com/Open-CMSIS-Pack/devtools/pull/2530

## Changes
- Ensures compile macro dumps use the appropriate C or C++ preprocessing mode and compiler options.
- Pass `-xc` or `-xc++`, or `--c++` in case of IAR, plus language-specific and common compiler options to macro dump commands.
- Update unit coverage and generated CMake reference fixtures across supported toolchains.
- Generating separate `compile_macros_c.h` and `compile_macros_cxx.h` files for enabled C and C++ languages has several advantages over a single-file approach:
  - Each file is a direct dump from the compiler preprocessor for one language mode.
  - No merging, comparison, filtering, or semantic transformation is required.
  - C and C++ differences are preserved exactly as reported by the compiler.
  - The consumer selects the file through the translation unit’s language, not through logic inside the generated file.
  - It avoids relying on any macro/define before the generated macro environment has been established.
  
## Risk / limitations
- The previous shared `compile_macros.h` output is replaced by language-specific files, which may affect consumers relying on that filename.
- Macro generation remains limited to C and C++.

## Checklist
- [x] 🤖 This change is covered by unit tests (if applicable).
- [x] 🤹 Manual testing has been performed (if necessary).
- [x] 🛡️ Security impacts have been considered (if relevant).
- [x] 📖 Documentation updates are complete (if required).
- [x] 🧠 Third-party dependencies and TPIP updated (if required).

### CODE REVIEW

File: pkg/maker/contextlists.go [hunk 1]
```diff
@@ -176,6 +176,48 @@ include("components.cmake")
...
+		}
+
+		options += "\nset(CPP_OPTIONS_" + language
+		if len(languageOption) > 0 {
+			options += " \"" + languageOption + "\""
+		}
+		for _, option := range miscOptions {
+			option = strings.ReplaceAll(c.AdjustRelativePath(option), "\"", "\\\"")
+			options += " \"" + option + "\""
+		}
```
[the remark this answers is not in the document]
Comment (brondani) [reply pkg/maker/contextlists.go:213]: Addressed in https://github.com/Open-CMSIS-Pack/cbuild2cmake/pull/503/commits/8f84492b1888c78e098327695209834aa882298e for all compilers. Preserve escaped quotes such as paths with spaces.

**Comment by soumeh01** [comment]
@grasci-arm Review this change

**Comment by grasci-arm** [comment]
Plan:

Development plan for PR #503:

1. Confirm dependency readiness: verify Open-CMSIS-Pack/devtools#2530 has landed or that this PR is tested against a cbuild output containing the expected language-specific misc option fields (`misc.C`, `misc.CPP`, `misc.C-CPP`) and toolchain metadata.

2. Branch/base status: the PR branch is `generate-language-specific-compile` at `318e971`, with three commits ahead of `origin/main`; `origin/main` is four CI/dependency workflow commits ahead. The PR source changes are in `pkg/maker/contextlists.go`, `pkg/maker/contextlists_test.go`, and generated `test/data/solutions/**/ref/**/CMakeLists.txt` fixtures, so rebasing onto current `origin/main` should not require source-level conflict resolution.

3. Implementation focus: keep macro dump generation centralized in `Cbuild.PreprocessorOptions()` in `pkg/maker/contextlists.go`. For each enabled language, generate `COMPILE_MACROS_C` -> `compile_macros_c.h` and/or `COMPILE_MACROS_CXX` -> `compile_macros_cxx.h`; ignore non-C/CXX languages such as ASM. Include language mode flags (`-xc`, `-xc++`; for IAR, no C language flag and `--c++` for CXX) plus language-specific and common misc options.

4. CMake integration: ensure `CreateContextCMakeLists()` wires the generated macro files into the `database` target dependencies and emits a single `add_custom_command(OUTPUT ...)` containing one dump command per enabled C/CXX language. Confirm the previous shared `COMPILE_MACROS ${OUT_DIR}/compile_macros.h` path is fully removed from generated references.

5. Option handling: review the split/quoting behavior for misc options with spaces, quoted include paths, escaped quotes, and relative paths adjusted through `AdjustRelativePath()`. If any edge case is ambiguous, add a narrow unit test before changing the implementation.

6. Test coverage: keep and, if needed, extend `pkg/maker/contextlists_test.go` for mixed ASM/C/CXX, C-only, CXX-only, IAR behavior, common `C-CPP` options, and quoted path handling. Fixture coverage should continue to cover GCC, AC6, CLANG, and IAR contexts under `test/data/solutions/**/ref/**/CMakeLists.txt`.

7. Fixture regeneration/validation: regenerate reference CMake outputs using the repository’s normal generator flow, then review fixture diffs to confirm each context only gains the expected `CPP_OPTIONS_*`, `COMPILE_MACROS_*`, database dependency, and macro dump command changes.

8. Verification: run `gofmt -s -w` on touched Go files, `git diff --check`, `go test ./...`, and preferably `make test-all` in an environment with Go and golangci-lint installed. This container lacks `go`, so rely on CI or a dev machine for the full test run.

9. Compatibility check: because `compile_macros.h` is intentionally replaced, confirm downstream consumers for cmsis-toolbox#625 and vscode-cmsis-solution#335 select `compile_macros_c.h` or `compile_macros_cxx.h` based on translation-unit language, and note the filename change in release/consumer-facing documentation if required.

10. Final review gate: after dependency PR #2530 and CI are green, rebase or update the branch onto current `origin/main`, verify no generated fixtures drift, and request focused review from maintainers on IAR option handling and downstream consumer expectations.

Reply with a comment mentioning the service account and an approval phrase such as `approve`, `approved`, `proceed`, or `go ahead`, or add a thumbs-up reaction to this plan comment, before execution.

**Review by soumeh01 — APPROVED** [review]
LGTM
